hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+    bb_reachable(c, cfg, bb) <-- cfg_node(c, cfg),
+        bb_reachable(c, cfg, pred),
+        io_node(c, pred, pred_out, IO::Output),
+        in_wire_value(c, pred_out, IncomingPort::from(0), predicate),
+        for (tag, bb) in c.output_neighbours(*pred).enumerate(),
+        if predicate.supports_tag(tag);
+
+    // Where do the values "fed" along a control-flow edge come out?
+    relation _cfg_succ_dest(C, Node, Node, Node);
+    _cfg_succ_dest(c, cfg, blk, inp) <-- dfb_block(c, cfg, blk), io_node(c, blk, inp, IO::Input);
+    _cfg_succ_dest(c, cfg, exit, cfg) <-- cfg_node(c, cfg), if let Some(exit) = c.children(*cfg).nth(1);
+
+    // Inputs of CFG propagate to entry block
+    out_wire_value(c, i_node, OutgoingPort::from(p.index()), v) <--
+        cfg_node(c, cfg),
+        if let Some(entry) = c.children(*cfg).next(),
+        io_node(c, entry, i_node, IO::Input),
+        in_wire_value(c, cfg, p, v);
+
+    // Outputs of each reachable block propagated to successor block or (if exit block) then CFG itself
+    out_wire_value(c, dest, OutgoingPort::from(out_p), v) <--
+        dfb_block(c, cfg, pred),
+        bb_reachable(c, cfg, pred),
+        let df_block = c.get_optype(*pred).as_dataflow_block().unwrap(),
+        for (succ_n, succ) in c.output_neighbours(*pred).enumerate(),
+        io_node(c, pred, out_n, IO::Output),
+        _cfg_succ_dest(c, cfg, succ, dest),
+        node_in_value_row(c, out_n, out_in_row),
+        if let Some(fields) = out_in_row.unpack_first(succ_n, df_block.sum_rows.get(succ_n).unwrap().len()),
+        for (out_p, v) in fields.enumerate();
+
+    // Call
+    relation func_call(C, Node, Node);
+    func_call(c, call, func_defn) <--
+        node(c, call),
+        if c.get_optype(*call).is_call(),
+        if let Some(func_defn) = c.static_source(*call);
+
+    out_wire_value(c, inp, OutgoingPort::from(p.index()), v) <--
+        func_call(c, call, func),
+        io_node(c, func, inp, IO::Input),
+        in_wire_value(c, call, p, v);
+
+    out_wire_value(c, call, OutgoingPort::from(p.index()), v) <--
+        func_call(c, call, func),
+        io_node(c, func, outp, IO::Output),
+        in_wire_value(c, outp, p, v);
+}
+
+fn propagate_leaf_op<V: AbstractValue>(
+    c: &impl DFContext<V>,
+    n: Node,
+    ins: &[PV<V>],
+    num_outs: usize,
+) -> Option<ValueRow<V>> {
+    match c.get_optype(n) {
+        // Handle basics here. I guess (given the current interface) we could allow
+        // DFContext to handle these but at the least we'd want these impls to be
+        // easily available for reuse.
+        op if op.cast::<MakeTuple>().is_some() => Some(ValueRow::from_iter([PV::new_variant(
+            0,
+            ins.iter().cloned(),
+        )])),
+        op if op.cast::<UnpackTuple>().is_some() => {
+            let elem_tys = op.cast::<UnpackTuple>().unwrap().0;
+            let [tup] = ins.iter().collect::<Vec<_>>().try_into().unwrap();
```

Comment:
This could use `Itertools::exactly_one` to avoid allocating a vector.

Reply:
Done, ta